File: algorithms/sorting/counting-sort/counting-sort.mjs
export function countingSort(numbers) {
  if (numbers.some((value) => !Number.isInteger(value))) {
    throw new TypeError("Counting Sort requires integer values");
  }
  if (numbers.length === 0) return [];

  const minimum = Math.min(...numbers);
  const maximum = Math.max(...numbers);
  const counts = Array(maximum - minimum + 1).fill(0);

  for (const value of numbers) counts[value - minimum] += 1;
  for (let index = 1; index < counts.length; index += 1) {
    counts[index] += counts[index - 1];
  }

  const output = Array(numbers.length);
  for (let index = numbers.length - 1; index >= 0; index -= 1) {
    const value = numbers[index];
    counts[value - minimum] -= 1;
    output[counts[value - minimum]] = value;
  }
  return output;
}

File: algorithms/sorting/counting-sort/counting_sort.py
def counting_sort(numbers):
    if any(not isinstance(value, int) for value in numbers):
        raise TypeError("Counting Sort requires integer values")
    if not numbers:
        return []

    minimum = min(numbers)
    maximum = max(numbers)
    counts = [0] * (maximum - minimum + 1)

    for value in numbers:
        counts[value - minimum] += 1
    for index in range(1, len(counts)):
        counts[index] += counts[index - 1]

    output = [0] * len(numbers)
    for value in reversed(numbers):
        counts[value - minimum] -= 1
        output[counts[value - minimum]] = value
    return output
